### ISSUE TITLE
incorrect `FromEnv` predicates around supertraits

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#49953]
The predicates for supertraits look wrong, as you can see from this example:

```rust
#![feature(rustc_attrs)]
#![allow(dead_code)]

trait Foo { }

#[rustc_dump_program_clauses] //~ ERROR program clause dump
trait Bar: Foo { }
```

I see this:

```
error: program clause dump
  --> lower_env1.rs:16:1
   |
16 | #[rustc_dump_program_clauses] //~ ERROR program clause dump
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: Implemented(Self: Bar) :- FromEnv(Self: Bar).
   = note: FromEnv(Self: Bar) :- FromEnv(Self: Bar).
```

But I expect `FromEnv(Self: Foo) :- FromEnv(Self: Bar)`.

cc @rust-lang/wg-traits

**Comment by scalexm** [comment]
Could it be because of the fact that for the moment, we unconditionally skip the first predicate (because it’s supposed to be `Self: Bar`)? Maybe in that case we are actually skipping the `Self: Foo` predicate?

**Comment by scalexm** [comment]
cc @tmandry

**Comment by tmandry** [comment]
Yeah I strongly suspect you are correct @scalexm. It looks like that hack doesn't work in all cases. I'm working to remove the `Self: Foo` from `predicates_of` so the hack isn't necessary, but if we want a fix now, it should be possible to run through all the predicates and filter out the one that matches `Self: Foo`. I'll try to submit one of those fixes tomorrow, if possible.
